### PR TITLE
Push to hub mixins that do not leverage `git`

### PR DESCRIPTION
**Pull request by LysandreJik**
This draft PR proposes a refactor of both mixins, `ModelHubMixin` and `push_to_hub_keras`, so that they do not leverage `git` under the hood. It is the continuation of https://github.com/huggingface/huggingface_hub/pull/321.

This PR was done to present how it would behave. It was not done with backward-compatibility in mind, but instead by thinking of the best API, therefore with lean signatures. I recommend reading the files themselves rather than the diff as they are drastically simpler:

If we choose to go this way, a standard backward compatibility approach would be implemented.

<table>
<tr>
<td>PyTorch</td>
<td>Keras</td>
</tr>
<tr>
<td>
Relevant code: <a href='https://github.com/LysandreJik/huggingface_hub/blob/6389eaa0d519e3121ba0a55cf6bacb19d520464b/src/huggingface_hub/hub_mixin.py#L218-L286'>ModelHubMixin</a>
</td>
<td>
Relevant code: <a href='https://github.com/LysandreJik/huggingface_hub/blob/6389eaa0d519e3121ba0a55cf6bacb19d520464b/src/huggingface_hub/keras_mixin.py#L271-L369'>Keras</a>
</td>
</tr>
<tr>
<td>

```py
def push_to_hub(
    self,
    repo_id: str,
    *,
    commit_message: Optional[str] = "Add model",
    private: Optional[bool] = None,
    api_endpoint: Optional[str] = None,
    token: Optional[str] = None,
    branch: Optional[str] = None,
    config: Optional[dict] = None,
) -> str:
```

</td>
<td>

```py
def push_to_hub_keras(
    model,
    repo_id: str,
    *,
    log_dir: Optional[str] = None,
    commit_message: Optional[str] = "Add model",
    private: Optional[bool] = None,
    api_endpoint: Optional[str] = None,
    token: Optional[str] = True,
    branch: Optional[str] = None,
    config: Optional[dict] = None,
    include_optimizer: Optional[bool] = False,
    task_name: Optional[str] = None,
    plot_model: Optional[bool] = True,
    **model_save_kwargs,
):
```

</td>
</tr>
<tr>
<td>

```py
import os.path
from typing import List

import torch

from huggingface_hub import ModelHubMixin
from torch import nn


class MyModel(nn.Module, ModelHubMixin):
    def __init__(self):
        super().__init__()

        self.ln_1 = nn.Linear(10, 100)
        self.ln_2 = nn.Linear(100, 10)

    def __call__(self, *args, **kwargs):
        return self.ln_1(self.ln_2(*args, **kwargs))

    def _save_pretrained(self, save_directory) -> List[str]:
        path = os.path.join(save_directory, 'pytorch_model.bin')
        torch.save(self, path)

        return [path]


MyModel().push_to_hub(
    'Jikiwa/my-new-model', 
    commit_message="Adding a config.", 
    config={'parameter': 2}
)
``` 

</td>
<td>

```py
import tensorflow as tf

from huggingface_hub import push_to_hub_keras


class MyModel(tf.keras.Model):
    def __init__(self, **kwargs):
        super().__init__()
        self.l1 = tf.keras.layers.Dense(2, activation="relu")
        dummy_batch_size = input_dim = 2
        self.dummy_inputs = tf.ones([dummy_batch_size, input_dim])

    def call(self, x):
        return self.l1(x)


model = MyModel()
model(tf.ones([2, 2]))
push_to_hub_keras(
    model, 
    'Jikiwa/my-new-keras-model', 
    commit_message="Committing model number 2!"
)
```

</td>
</tr>
</table>

cc @osanseviero @NielsRogge @nateraw @merveenoyan

### CODE REVIEW

**Comment by adrinjalali** [comment]
Merging https://github.com/huggingface/huggingface_hub/pull/849 should reduce the diff here.

**Comment by merveenoyan** [comment]
I love the refactor with HF API! I can open a PR for tests if you want @LysandreJik

**Comment by LysandreJik** [comment]
Thank you for the proposal @merveenoyan, I'll likely proceed to a backward compatibility implementation + test update next week. I'll let you know if I can use assistance on any of it!

File: src/huggingface_hub/hub_mixin.py
Comment (fcakyon) [inline src/huggingface_hub/hub_mixin.py:279]: In the docs it says `upload_file` should be used when the file size is less than 10MB: https://huggingface.co/docs/hub/adding-a-model

Can this new implementation handle weights files having 100-200MB size?

Comment (LysandreJik) [reply src/huggingface_hub/hub_mixin.py:279]: This implementation can handle weights files up to 30GB!

Comment (LysandreJik) [reply src/huggingface_hub/hub_mixin.py:279]: You're correct that the documentation should likely be updated to reflect that.

File: src/huggingface_hub/hf_api.py
Comment (nateraw) [inline src/huggingface_hub/hf_api.py:666]: Should we also add `ValueError` on file size being greater than 30GB?

File: src/huggingface_hub/hf_api.py
Comment (nateraw) [inline src/huggingface_hub/hf_api.py:1840]: Should we offer any additional guidance here? As a user, you may find yourself sort of out of luck if you've already planned to use the mixin for this (if that's where it ends up being surfaced)

File: src/huggingface_hub/hf_api.py
Comment (nateraw) [inline src/huggingface_hub/hf_api.py:1883]: Not sure what this response looks like in the case of an error...should this be raising something as well?

File: src/huggingface_hub/hub_mixin.py
Comment (nateraw) [inline src/huggingface_hub/hub_mixin.py:248]: In regards to earlier comment about warning user about files larger than 30GB, maybe that can be put here more explicitly? Not sure which place is better (or maybe both places?)

File: src/huggingface_hub/hub_mixin.py
Comment (nateraw) [inline src/huggingface_hub/hub_mixin.py:333]: The parent class's `_save_pretrained` doesn't specify returning anything, so adding this here throws me off a bit. But it seems like a nice idea, especially if folks are going to be overriding this. Should we maybe document this a bit to explain what's going on?

File: src/huggingface_hub/keras_mixin.py
Comment (nateraw) [inline src/huggingface_hub/keras_mixin.py:310]: should this be `huggingface-cli login`?

**Review by nateraw — COMMENTED** [review]
I'm excited to move towards this type of logic 😄  - here's a few comments while its still WIP.

**Comment by LysandreJik** [comment]
This is on hold as another, more robust mechanism to upload files is in the making on the backend side.

**Comment by fcakyon** [comment]
> This is on hold as another, more robust mechanism to upload files is in the making on the backend side.

🤔 waiting for the new upload mechanism to integrate push to hub functionality to my packages!

**Comment by LysandreJik** [comment]
Glad to hear it @fcakyon, we will let you know.

File: src/huggingface_hub/hub_mixin.py
Comment (Wauplin) [reply src/huggingface_hub/hub_mixin.py:441]: In the end we didn't need this return so best thing for now is to remove it.

**Comment by Wauplin** [comment]
@SBrandeis @julien-c @nateraw Work on `push_to_hub` feature for mixins (using HTTP endpoints) is now ready for review. 

Changes have first been reviewed by @LysandreJik in https://github.com/LysandreJik/huggingface_hub/pull/8 before merging into his branch (after some feedback/changes, see discussions). Main changes compared to before is handling backward compatibility with deprecated arguments and some tests and docstring.

**Comment by LysandreJik** [comment]
Pinging @NielsRogge as he wants to play with it and to help advertise it after this is merged!

**Comment by fcakyon** [comment]
I was using PytorchModelMixin class for my upcoming open source package, is this PR going to affect the API? If yes, how should I migrate to the upcoming API?

**Comment by Wauplin** [comment]
> I was using PytorchModelMixin class for my upcoming open source package, is this PR going to affect the API? If yes, how should I migrate to the upcoming API?

Hi @fcakyon, once PR is merged and `huggingface_hub` v0.9 is released you should be able to push to hub using the non-git mixins the same way as before but using different arguments:

```py
model.push_to_hub(repo_id="mymodel", commit_message="model-1")
```

**Note:** previous git-based implementation still works and is expected to be supported until v0.12. Using deprecated arguments will prompt you some `FutureWarnings`.

---
Full list of arguments in new version will be:
```py
def push_to_hub(
    self,
    repo_id: str,
    *,
    commit_message: Optional[str] = "Add model",
    private: Optional[bool] = None,
    api_endpoint: Optional[str] = None,
    token: Optional[str] = None,
    branch: Optional[str] = None,
    create_pr: Optional[bool] = None,
    config: Optional[dict] = None,
)
 ```

File: setup.py
Comment (osanseviero) [inline setup.py:38]: Not sure about this. Maybe extras["integrations"] would be better name but not convinced

File: src/huggingface_hub/hub_mixin.py
Comment (osanseviero) [inline src/huggingface_hub/hub_mixin.py:46]: nit
```suggestion
                Specify config (must be dict) in case you want to save
```

File: src/huggingface_hub/hub_mixin.py
Comment (osanseviero) [inline src/huggingface_hub/hub_mixin.py:85]: Should we have a deprecation warning here or do we think with `push_to_hub` warning it's enough?

File: src/huggingface_hub/hub_mixin.py
Comment (osanseviero) [inline src/huggingface_hub/hub_mixin.py:287]:
```suggestion
                The repository to which the model will be pushed, for example:
                `"username/repo_name"`
```

File: src/huggingface_hub/hub_mixin.py
Comment (osanseviero) [inline src/huggingface_hub/hub_mixin.py:312]:
```suggestion
        # If the repo id is set, it means we use the new version using HTTP endpoint
        # (introduced in v0.9).
```

File: src/huggingface_hub/hub_mixin.py
Comment (osanseviero) [inline src/huggingface_hub/hub_mixin.py:331]: No preference, but this is not needed. (maybe being explicit is better as you do here)

File: src/huggingface_hub/hub_mixin.py
Comment (osanseviero) [inline src/huggingface_hub/hub_mixin.py:341]:
```suggestion
        # If the repo id is None, it means we use the deprecated version using Git
        # TODO: remove code between here and `return repo.git_push()` in release 0.12
```

File: src/huggingface_hub/keras_mixin.py
Comment (osanseviero) [inline src/huggingface_hub/keras_mixin.py:14]: Is importing `ModelHubMixin` needed?

**Comment by fcakyon** [comment]
@Wauplin thanks a lot! So by setting `repo_id=organization_name/model_name` I will be able to create a fresh model or update an existing model in the hub, right?

**Comment by Wauplin** [comment]
> @Wauplin thanks a lot! So by setting `repo_id=organization_name/model_name` I will be able to create a fresh model or update an existing model in the hub, right?

Exactly 👍

File: tests/test_hubmixin.py
Comment (osanseviero) [inline tests/test_hubmixin.py:106]: If it's not much I would consider splitting the tests as they testing different behaviours

File: tests/test_hubmixin.py
Comment (osanseviero) [inline tests/test_hubmixin.py:146]: Shouldn't these raise warnings? Why not use `expect_deprecation` for these? (also a bonus point for splitting the tests)

File: tests/test_keras_integration.py
Comment (osanseviero) [inline tests/test_keras_integration.py:439]: Doesn't this need `expect_deprecation`?

File: tests/testing_utils.py
Comment (osanseviero) [inline tests/testing_utils.py:236]: Very cool!

**Review by osanseviero — COMMENTED** [review]
Thanks for the PR! :partying_face: 

I did a high-level pass and left some very minor comments, but it is looking great! Having some docs/example on the new usage would be nice.

File: src/huggingface_hub/hub_mixin.py
Comment (julien-c) [inline src/huggingface_hub/hub_mixin.py:332]: this line looks weird, wondering if we can remove it?

Comment (osanseviero) [reply src/huggingface_hub/hub_mixin.py:332]: I don't think so as `path_in_repo` is not optional, but I think it should. The API looks weird indeed, see https://github.com/huggingface/huggingface_sb3/blob/main/huggingface_sb3/push_to_hub.py#L389-L395 as an example usage

Comment (julien-c) [reply src/huggingface_hub/hub_mixin.py:332]: yep sounds good to me to do in a subsequent PR!

File: src/huggingface_hub/hub_mixin.py
Comment (Wauplin) [reply src/huggingface_hub/hub_mixin.py:85]: I can deprecate the argument in `save_pretrained` as well but I thought that since `save_pretrained` is just forwarding the kwargs to `push_to_hub` it was fine to warn only once.

But I get your point so if we really want to deprecate in `save_pretrained` as well, I really don't mind adding the decorator here as well.

File: src/huggingface_hub/keras_mixin.py
Comment (Wauplin) [reply src/huggingface_hub/keras_mixin.py:14]: I did not add it myself (it's just an isort refactoring here). But it seems `ModelHubMixin` is use in the `KerasModelHubMixin` class. Is this a deprecated class ?

If you think yes, I can open an issue to deprecate it.

File: tests/test_keras_integration.py
Comment (Wauplin) [reply tests/test_keras_integration.py:439]: Yes I forgot this one, thanks 👍

File: tests/test_hubmixin.py
Comment (Wauplin) [reply tests/test_hubmixin.py]: @osanseviero Hmm, yes and no. Here I'm testing how `save_pretrained` propagate keyword arguments to `push_to_hub` in 5 different cases. Since I did not want to test `push_to_hub` itself I just used a mock for it to check which arguments are passed. Since `push_to_hub` itself is never run in the test, the deprecation warnings are not shown. To make it more explicit, I renamed `model` to `mocked_model`.

About splitting tests, I get the point what checking different cases but I didn't know how to do it. A basic way would be to duplicate init code for each test but I'm not a huge fan:

```py
def test_case_xxx(self):
    # next 6 lines for init
    REPO_NAME = repo_name("save")
    save_directory = f"{WORKING_REPO_DIR}/{REPO_NAME}"
    config = {"hello": "world"}
    mocked_model = DummyModel()
    mocked_model.push_to_hub = Mock()
    mocked_model._save_pretrained = Mock()

    mocked_model.save_pretrained(...) # <- actual code that changes for each case
    mocked_model.push_to_hub.assert_called_with(...) # <- actual code that changes for each case
```
I could also group the first 6 lines in a separate function but it would make the test less readable.

In the end, that's why I kept all the cases together (also the fact that they don't have any external side effect since it's all mocked) but I'd be happy to change if you have an opinion/suggestion about it.

File: tests/test_hubmixin.py
Comment (Wauplin) [reply tests/test_hubmixin.py]: See comment below.

File: tests/test_hubmixin.py
Comment (Wauplin) [reply tests/test_hubmixin.py]: TL;DR: I am open to suggestion on how to make this clean 😇

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py:2096]: Oups, good to know. Sorry, next time I'll not move it then

File: src/huggingface_hub/hub_mixin.py
Comment (Wauplin) [reply src/huggingface_hub/hub_mixin.py:332]: Issue created here https://github.com/huggingface/huggingface_hub/issues/971 to track that. I resolve this conversation.

File: setup.py
Comment (julien-c) [reply setup.py]: maybe `extras["frameworks"]` (though it's somewhat vague)

File: src/huggingface_hub/hub_mixin.py
Comment (osanseviero) [reply src/huggingface_hub/hub_mixin.py:85]: No strong opinion, up to you!

Comment (Wauplin) [reply src/huggingface_hub/hub_mixin.py:85]: Ok, I'll keep it as it is then.

File: src/huggingface_hub/hf_api.py
Comment (LysandreJik) [reply src/huggingface_hub/hf_api.py:659]: I wonder why all those changes happened here, they seem unrelated to the PR right?

File: src/huggingface_hub/hub_mixin.py
Comment (LysandreJik) [inline src/huggingface_hub/hub_mixin.py:422]: Is this the only place where this needs to be changed? There might be some documentation in `docs/` that also need to be updated to reflect the new API

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py:659]: No idea, I will revert that 😬

File: src/huggingface_hub/hub_mixin.py
Comment (Wauplin) [reply src/huggingface_hub/hub_mixin.py:422]: Actually I went though the documentation and it doesn't seem to be mentioned elsewhere.
In fact, `push_to_hub` mixins possibilities are documented [on this page](https://huggingface.co/docs/huggingface_hub/main/en/package_reference/mixins) which is more a package reference page.

File: src/huggingface_hub/hub_mixin.py
Comment (Wauplin) [reply src/huggingface_hub/hub_mixin.py]: Will be discussed in a separate issue https://github.com/huggingface/huggingface_hub/issues/995.

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py]: Will be discussed in a separate issue https://github.com/huggingface/huggingface_hub/issues/995.

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py]: Will be discussed in a separate issue https://github.com/huggingface/huggingface_hub/issues/995.

File: src/huggingface_hub/hub_mixin.py
Comment (Wauplin) [reply src/huggingface_hub/hub_mixin.py:301]: Will be discussed in a separate issue https://github.com/huggingface/huggingface_hub/issues/995.

**Review by Wauplin — APPROVED** [review]

**Comment by LysandreJik** [comment]
Big effort! Let's merge this, thanks for the help and all the work improving & wrapping it up @Wauplin!

**Comment by Wauplin** [comment]
Many thanks for the help and support to finally make it @LysandreJik ! 😀

**Comment by fcakyon** [comment]
Great PR :100:

**Pull request merged**
